### PR TITLE
[AGMS-1075] Reuse functions from backend tests in prepare script

### PR DESCRIPTION
**Pull request by jhellar**
## Motivation
https://issues.jboss.org/browse/AGMS-1075

### CODE REVIEW

File: device/scripts/prepare.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+const getMobileClientCr = require('../../common/templates/mobile-client');
+const getSyncConfigMap = require('../../common/templates/data-sync');
+const getKeycloakRealmCr = require('../../common/templates/keycloak-realm');
+const getAndroidVariantCr = require('../../common/templates/android-variant');
+const getMssAppCr = require('../../common/templates/mss-app');
+const { waitFor, randomString } = require('../../common/util/utils');
+
+const appName = 'test-suite-app';
+const syncNamespacePrefix = 'test-suite-sync-';
+const syncNamespace = syncNamespacePrefix + randomString();
```
Comment (b1zzu) [inline device/scripts/prepare.js]: Why don't using a fixed name for the namespaces like `device-test-suite-sync`? In any case, all namespaces starting with `test-suite-sync-` will be deleted so I don't get why adding a random part to it

File: backend/scripts/prepare.sh [hunk 2]
```diff
@@ -1,20 +1,7 @@
 SYNC_NAMESPACE_PREFIX="test"
```
Comment (b1zzu) [inline backend/scripts/prepare.sh]: Can we also convert this prepare.sh in javascript, or are you doing it in a sperate PR

File: package.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "test-suite",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (b1zzu) [inline package.json]: We are not deploying this project to npm so in my opinion information like `description`, `main`, `repository`, `keywords`, `author`, `license`, `bugs`, `hompeage`, and probably also `version` are useless

File: showcase/scripts/prepare.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    app = await resource(TYPE.MOBILE_APP, ACTION.GET, mobileApp.metadata.name);
+    return app.status.services.length === 2;
+  }, 30 * 1000);
+
+  // get mobile-services.json
+  console.log('Getting mobile-services.json...');
+  fs.writeFileSync(path.resolve(__dirname, '../mobile-services.json'), JSON.stringify(app.status, null, 2));
+
+  console.log('Setup successful');
+})();
```
Comment (b1zzu) [inline showcase/scripts/prepare.js]: There is a lot of duplicated code between this script and `device/scripts/prepare.js`, maybe we can create a common function with an `options` args to requesting different capabilities, like `sync`, `push`, ... or an object with an init method for the Openshift credentials and creating the mdc app, and a sperate method for each service that we can bind, and a last method to export the mobile-services.json, wdyt?

File: device/scripts/prepare.js (hunk 1 above)
Comment (b1zzu) [reply device/scripts/prepare.js]: Also if the `appName` is unique for each suite we could run the **device** and **showcase** test suite in parallel

Comment (jhellar) [reply device/scripts/prepare.js]: Reason for the random part is that for some reason in script when I run the delete command for the namespace then create command, it complains that the namespace still exists. I've tried to add sleep after deletion, but no matter how long it waits it complains every time. This is just a workaround.

Comment (jhellar) [reply device/scripts/prepare.js]: Makes sense to have different appName for the tests, will fix it.

File: backend/scripts/prepare.sh (hunk 2 above)
Comment (jhellar) [reply backend/scripts/prepare.sh]: Not sure if it is worth doing it, it only deploys showcase server and in the `js` code, it is also done by executing svcat command.

File: package.json (hunk 3 above)
Comment (jhellar) [reply package.json]: ok, will fix this

File: showcase/scripts/prepare.js (hunk 4 above)
Comment (jhellar) [reply showcase/scripts/prepare.js]: That makes sense, will look at it.

File: device/scripts/prepare.js (hunk 1 above)
Comment (b1zzu) [reply device/scripts/prepare.js]: > Reason for the random part is that for some reason in script when I run the delete command for the namespace then create command, it complains that the namespace still exists. I've tried to add sleep after deletion, but no matter how long it waits it complains every time. This is just a workaround.

This a good reason, can we add a short comment to the code to not forget it?

File: backend/scripts/prepare.sh (hunk 2 above)
Comment (b1zzu) [reply backend/scripts/prepare.sh]: I'm still thinking that having it in Javascript for all our suits will simplify the maintenance process in the future

File: device/scripts/prepare.js (hunk 1 above)
Comment (b1zzu) [reply device/scripts/prepare.js]: > Reason for the random part is that for some reason in script when I run the delete command for the namespace then create command, it complains that the namespace still exists. I've tried to add sleep after deletion, but no matter how long it waits it complains every time. This is just a workaround.

This a good reason, can we add a short comment to the code to not forget it?

File: backend/test/app-deletion.js [hunk 5]
```diff
@@ -41,7 +41,7 @@ describe('App deletion', async function() {
   };
 
   before('init kube client', async function() {
-    await initKubeClient(process.env.MDC_NAMESPACE);
+    await init(process.env.MDC_NAMESPACE);
```
Comment (psturc) [inline backend/test/app-deletion.js]:
```suggestion
    await init();
```

File: backend/test/bindings.js [hunk 6]
```diff
@@ -41,7 +41,7 @@ describe('Bindings', async function() {
   };
 
   before('init kube client', async function() {
-    await initKubeClient(process.env.MDC_NAMESPACE);
+    await init(process.env.MDC_NAMESPACE);
```
Comment (psturc) [inline backend/test/bindings.js]:
```suggestion
    await init();
```

Comment (b1zzu) [reply backend/test/bindings.js]: How did you suggest changes?

**Comment by b1zzu** [comment]
@jhellar we should document in the **device** and in the **showcase** test suite that before `prepare.js` we have to run `npm install` on the parent directory

**Comment by b1zzu** [comment]
@jhellar I have a couple of warnings when I run `prepare.js`
```
kubernetes-client deprecated require('kubernetes-client').config, use require('kubernetes-client/backends/request').config. ../node_modules/openshift-rest-client/lib/openshift-rest-client.js:25:17
kubernetes-client deprecated require('kubernetes-client').config, use require('kubernetes-client/backends/request').config. ../node_modules/openshift-rest-client/index.js:23:39
kubernetes-client deprecated fromKubeconfig see https://github.com/godaddy/kubernetes-client/blob/master/merging-with-kubernetes.md#request-kubeconfig- ../node_modules/openshift-rest-client/lib/openshift-rest-client.js:88:60
kubernetes-client deprecated Client({ config }), see https://github.com/godaddy/kubernetes-client/blob/master/merging-with-kubernetes.md ../node_modules/openshift-rest-client/lib/openshift-rest-client.js:88:18
kubernetes-client deprecated Request() without a .kubeconfig option, see https://github.com/godaddy/kubernetes-client/blob/master/merging-with-kubernetes.md ../node_modules/kubernetes-client/lib/swagger-client.js:200:19
```
should we do something for them?

File: backend/test/bindings.js (hunk 6 above)
Comment (psturc) [reply backend/test/bindings.js]: Screenshot 2019-10-09 at 13 38 24

File: device/scripts/prepare.js (hunk 1 above)
Comment (jhellar) [reply device/scripts/prepare.js]: ok, will add a comment.

File: device/scripts/prepare.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  console.log('Creating push application...');
+  const pushApp = await createPushApp(appName);
+
+  // bind with push - android
+  console.log('Binding with push - android...');
+  const androidVariantCr = getAndroidVariantCr(
+    mobileApp.metadata.name,
+    mobileApp.metadata.uid,
+    pushApp.status.pushApplicationId,
+    process.env.FIREBASE_SERVER_KEY
```
Comment (b1zzu) [inline device/scripts/prepare.js]: Is FIREBASE_SENDER_ID optional? Should we remove it from the doc?

**Comment by jhellar** [comment]
@b1zzu the warnings are probably from openshift-rest-client, seems that it uses kubernetes-client internally, but using deprecated stuff.

File: device/scripts/prepare.js (hunk 7 above)
Comment (jhellar) [reply device/scripts/prepare.js]: It is still used in device tests: https://github.com/aerogear/test-suite/blob/master/device/test/push/index.ts#L22

**Comment by jhellar** [comment]
@b1zzu for the npm install in root directory, that also makes sense, will add it

File: device/scripts/prepare.js (hunk 7 above)
Comment (b1zzu) [reply device/scripts/prepare.js]: But maybe is not needed also there, I will try without it

**Comment by jhellar** [comment]
Rebased on master

**Comment by jhellar** [comment]
thanks @psturc missed that

**Comment by psturc** [comment]
> @b1zzu the warnings are probably from openshift-rest-client, seems that it uses kubernetes-client internally, but using deprecated stuff.

https://github.com/nodeshift/openshift-rest-client/pull/171

File: backend/scripts/prepare.sh (hunk 2 above)
Comment (jhellar) [reply backend/scripts/prepare.sh]: ok, should be fixed now

**Comment by jhellar** [comment]
@psturc awesome, no longer see the warnings with this change

**Comment by b1zzu** [comment]
@jhellar I have an issue with the `device/scripts/prepare.js`
```
./scripts/prepare.js          
kubernetes-client deprecated require('kubernetes-client').config, use require('kubernetes-client/backends/request').config. ../node_modules/openshift-rest-client/lib/openshift-rest-client.js:25:17
kubernetes-client deprecated require('kubernetes-client').config, use require('kubernetes-client/backends/request').config. ../node_modules/openshift-rest-client/index.js:23:39
kubernetes-client deprecated fromKubeconfig see https://github.com/godaddy/kubernetes-client/blob/master/merging-with-kubernetes.md#request-kubeconfig- ../node_modules/openshift-rest-client/lib/openshift-rest-client.js:88:60
kubernetes-client deprecated Client({ config }), see https://github.com/godaddy/kubernetes-client/blob/master/merging-with-kubernetes.md ../node_modules/openshift-rest-client/lib/openshift-rest-client.js:88:18
kubernetes-client deprecated Request() without a .kubeconfig option, see https://github.com/godaddy/kubernetes-client/blob/master/merging-with-kubernetes.md ../node_modules/kubernetes-client/lib/swagger-client.js:200:19
Recreating mobile app...
Redeploying showcase server...
Binding with services...
(node:485) UnhandledPromiseRejectionWarning: Error: IOSVariant.push.aerogear.org "device-test-suite-ios-ups-variant" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"push.aerogear.org/v1alpha1", "kind":"IOSVariant", "metadata":map[string]interface {}{"generation":1, "uid":"6b06b4f9-eb60-11e9-8670-0e70c7b29082", "selfLink":"", "clusterName":"", "name":"device-test-suite-ios-ups-variant", "creationTimestamp":"2019-10-10T13:18:15Z", "ownerReferences":[]interface {}{map[string]interface {}{"kind":"MobileClient", "name":"device-test-suite", "uid":"3992ea98-eb60-11e9-8670-0e70c7b29082", "blockOwnerDeletion":false, "apiVersion":"mdc.aerogear.org/v1alpha1"}}, "namespace":"mobile-developer-console"}, "spec":map[string]interface {}{"description":"UPS iOS Variant", "production":false, "pushApplicationId":"61ab2951-1348-46aa-b582-dee18b4113e4"}}: validation failure list:
spec.certificate in body is required
spec.passphrase in body is required
    at /home/dbizzarr/Projects/github.com/aerogear/test-suite/node_modules/kubernetes-client/backends/request/client.js:225:25
    at Request._callback (/home/dbizzarr/Projects/github.com/aerogear/test-suite/node_modules/kubernetes-client/backends/request/client.js:162:14)
    at Request.self.callback (/home/dbizzarr/Projects/github.com/aerogear/test-suite/node_modules/request/request.js:185:22)
    at Request.emit (events.js:209:13)
    at Request.<anonymous> (/home/dbizzarr/Projects/github.com/aerogear/test-suite/node_modules/request/request.js:1161:10)
    at Request.emit (events.js:209:13)
    at IncomingMessage.<anonymous> (/home/dbizzarr/Projects/github.com/aerogear/test-suite/node_modules/request/request.js:1083:12)
    at Object.onceWrapper (events.js:298:28)
    at IncomingMessage.emit (events.js:214:15)
    at endReadableNT (_stream_readable.js:1178:12)
(node:485) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:485) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
I think we should split android and ios and only bind android for the device tests

File: device/scripts/prepare.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  console.log('Redeploying showcase server...');
+  await redeployShowcase(appName);
+  
+  console.log('Binding with services...');
+  await bind(app, [
+    BINDING.DATA_SYNC,
+    BINDING.KEYCLOAK,
+    BINDING.MSS,
+    BINDING.PUSH_ANDROID,
+    BINDING.PUSH_IOS
```
Comment (b1zzu) [inline device/scripts/prepare.js]:
```suggestion
```

File: device/README.md [hunk 9]
```diff
@@ -54,7 +54,10 @@ export OPENSHIFT_HOST=<OPENSHIFT_HOST_URL>
 export OPENSHIFT_USER=<EVALS_USERNAME>
```
Comment (b1zzu) [inline device/README.md]: There is no need to set `OPENSHIFT_HOST` anymore

Comment (jhellar) [reply device/README.md]: ok, removed

**Comment by jhellar** [comment]
@b1zzu yes, the ios binding should not be there for device tests. Applied your change.

**Pull request merged**